pkg/resources/rbac.authorization.k8s.io/v1/clusterrole/validator.go

```diff
@@ -0,0 +1,92 @@
+package clusterrole
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/rbac/v1"
+
+	"github.com/rancher/webhook/pkg/admission"
+	objectsv1 "github.com/rancher/webhook/pkg/generated/objects/rbac.authorization.k8s.io/v1"
+	"github.com/rancher/webhook/pkg/resources/common"
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/trace"
+)
+
+const (
+	grOwnerLabel = "authz.management.cattle.io/gr-owner"
+)
+
+// Validator implements admission.ValidatingAdmissionHandler.
+type Validator struct {
+	admitter admitter
+}
+
+// NewValidator returns a new validator for roles.
+func NewValidator() *Validator {
+	return &Validator{
+		admitter: admitter{
+			clusterRoleOldAndNewFromRequest: objectsv1.ClusterRoleOldAndNewFromRequest,
+		},
+	}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group:    "rbac.authorization.k8s.io",
+		Version:  "v1",
+		Resource: "clusterroles",
+	}
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{
+		admissionregistrationv1.Update,
+	}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	webhook := admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.ClusterScope, v.Operations())
+	webhook.ObjectSelector = &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      grOwnerLabel,
+				Operator: metav1.LabelSelectorOpExists,
+			},
+		},
+	}
+	return []admissionregistrationv1.ValidatingWebhook{*webhook}
+}
+
+// Admitters returns the admitter objects used to validate roles.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+type clusterRoleOldAndNewFromRequest func(request *admissionv1.AdmissionRequest) (*v1.ClusterRole, *v1.ClusterRole, error)
+
+type admitter struct {
+	clusterRoleOldAndNewFromRequest clusterRoleOldAndNewFromRequest
```

Comment:
I appreciate that you are trying to make this more testable, but it's very possible to write tests for these objects without resorting to a change like this. [Example Code](https://github.com/rancher/webhook/blob/54bba7ed560004f40cd277aef921bdcd07802ea6/pkg/resources/management.cattle.io/v3/globalrole/validator.go#L84), [Example Test](https://github.com/rancher/webhook/blob/54bba7ed560004f40cd277aef921bdcd07802ea6/pkg/resources/management.cattle.io/v3/globalrole/validator_test.go#L760). Since we don't need this to support testing this function, I would prefer that you don't add an interface here.

Reply:
changed to follow the same approach as the GlobalRole test